py/objint: Add mp_obj_int_get_uint_checked() helper.

Jongy opened the pull request:
Can be used where mp_obj_int_get_checked() will overflow due
to the sign-bit solely.

Shouldn't affect code size, since it's unused at the moment.
Should any extmod/port use it, they'd "pay the price".

Addressed #2868, #2269 and possibly more.

stinos commented on py/objint.c:
```diff
@@ -363,6 +363,10 @@ mp_int_t mp_obj_int_get_checked(mp_const_obj_t self_in) {
 }
 
+mp_uint_t mp_obj_int_get_uint_checked(mp_const_obj_t self_in) {
+    return MP_OBJ_SMALL_INT_VALUE(self_in);
```
I'd probably lean to code reuse here and return mp_obj_int_get_checked so that if the implementation would change you only have to change it once. But I don't know if the compiler is smart enough to skip the extra call, nor if that matters.

Jongy replied:
I don't think we should trust the compiler to always skip the extra call. I could introduce another helper function/macro to be used by both functions, but I don't think it matters that much.

stinos commented on py/objint_longlong.c:
```diff
@@ -279,6 +279,11 @@ mp_int_t mp_obj_int_get_checked(mp_const_obj_t self_in) {
 }
 
+mp_uint_t mp_obj_int_get_uint_checked(mp_const_obj_t self_in) {
+    // TODO: Check overflow
```
Same remark as above. Bonus: you wouldn't have to repeat this comment.

Jongy replied:
Same reply, as well :)

stinos commented on py/objint_mpz.c:
```diff
@@ -411,6 +411,21 @@ mp_int_t mp_obj_int_get_checked(mp_const_obj_t self_in) {
+        if (mpz_as_uint_checked(&self->mpz, &value)) {
+            return value;
+        } else {
+            // overflow
```
That seems quite superfluous, given the next line

Jongy replied:
Hmm, right. Well, I just took it from `mp_obj_int_get_checked`. Will remove.

Jongy replied:
Not present in new function.

stinos commented:
> Shouldn't affect code size, since it's unused at the moment.

I'm not sure that is how it works. The code is there, and part of the public API so it makes it into the binary, no?

dpgeorge commented:
> I'm not sure that is how it works. The code is there, and part of the public API so it makes it into the binary, no?

If it's never used it won't make it into the binary, due to linker "garbage collection".

stinos commented:
I was actually thinking in terms of static/shared libraries, my mistake, that's not the standard way MicroPython gets used, so I wonder if that principle also applies there.

dpgeorge commented:
I think there are a few functions which are like this, part of the API but unused by the core (eg `mp_module_register`, kind of).

Jongy commented:
> I'm not sure that is how it works. The code is there, and part of the public API so it makes it into the binary, no?

As @dpgeorge also said, since MicroPython is compiled as a complete executable (and not as a shared library, a part of a bigger executable) and combined with the use of LTO / `-ffunction-sections` + `--gc-sections`, unused functions are cut off during the final linkage. Actually `mpz_as_uint_checked` which I call here is an example - it's not used (apart from `unix/coverage` builds) and it's still compiled whenever mpz is used.

Jongy commented:
Found another use-case for this function: `machine_mem_get_addr` calls `mp_obj_int_get_truncated`, basically because it can't call `mp_obj_int_get_checked` since there are legit addresses which don't fit in a signed word. So it calls `mp_obj_int_get_truncated` which works, but it gets messed up if you really use a value larger than an unsigned word. It should use `mp_obj_int_get_uint_checked`.

I can fix it here but it will cause users of `machine_mem` to automatically include `mp_obj_int_get_uint_checked` into the binary, so I'm not sure.

@dpgeorge What do you think about this function and the change?

Jongy commented:
> Maybe since the mp_obj_int_get_uint_checked implementation for mpz is basically the exact same as for signed int except for one function call the implementations can be merged to reduce code size, but it might be somewhat tricky due to the different return 

I made an attempt at it. I don't see any size diff (even when forcing the `uint` variant into the final binary), probably because the helper I made was inlined into them both. But I don't have time to verify that in the assembly now :(
Anyway I don't think it matters that much. I also think we better keep `mp_obj_int_get_checked` as it was, even if it makes the new helper function cost a bit more.

stinos commented:
> Anyway I don't think it matters that much. I also think we better keep `mp_obj_int_get_checked` as it was, even if it makes the new helper function cost a bit more.

Small incremental size additions could make for a large total.. But I mainly wonder why you think it's better to keep something, if changing it could be beneficial (don't know if it is though, didn't check)?

Jongy commented:
> Small incremental size additions could make for a large total.. But I mainly wonder why you think it's better to keep something, if changing it could be beneficial (don't know if it is though, didn't check)?

In the long run, after there are many users for `mp_obj_int_get_uint_checked`, it won't matter. Currently when first pulling this PR I think it's better to insert it as diff-less as possible and allow for individual call-sites to make the choice later.

After all, my intention here is merely to allow future users to choose this API, not to enforce it on current compilations (Which have worked fine *and will* work fine regardless of this PR)

stinos commented on py/objint_mpz.c:
```diff
@@ -411,6 +411,14 @@ mp_int_t mp_obj_int_get_checked(mp_const_obj_t self_in) {
+}
+
+mp_uint_t mp_obj_int_get_uint_checked(mp_const_obj_t self_in) {
+    return (mp_uint_t)_mp_obj_int_get_checked(self_in, (bool(*)(const mpz_t *, mp_int_t *))mpz_as_uint_checked);
```
Extra cast `(mp_uint_t)` shouldn't be needed since AFAIK C99 gurantees implicit conversions between arithmetic types. However, the casting of the function possibly needs some treatment. It's probably well-defined for all ports in use (didn't check though), but for the sake of never causing nasty problems I'd add an explicit compile-time check that the cast is correct. I think just checking that `sizeof(mp_int_t) == sizeof(mp_uint_t)` is sufficient.

Jongy replied:
Mhm, I'll add some compile-time test for this.

About the cast - IIRC I added it mainly for documentation. If I were to see this code w/o a cast, at a first glance I'd think "wow, this can implicitly sign underflow, that's bad". The cast clarifies the developer's intention.

stinos replied:
Fair enough but you didn't use the cast in any of the other places where it happens so i wasn't sure what was going in :)

Jongy replied:
Yeah technically `MP_OBJ_SMALL_INT_VALUE` is `mp_int_t` as well. But here the need for a cast was more evident, I think.

Anyway, resolving this, it's irrelevant now with the new code structure.

dpgeorge commented on py/objint_mpz.c:
```diff
@@ -396,13 +396,13 @@ mp_int_t mp_obj_int_get_truncated(mp_const_obj_t self_in) {
-mp_int_t mp_obj_int_get_checked(mp_const_obj_t self_in) {
+STATIC mp_int_t _mp_obj_int_get_checked(mp_const_obj_t self_in, bool (*as_int)(const mpz_t *i, mp_int_t *value)) {
     if (mp_obj_is_small_int(self_in)) {
         return MP_OBJ_SMALL_INT_VALUE(self_in);
```
What are the intended semantics of `mp_obj_int_get_uint_checked()` when the arg is a negative small-int?  Since it's "checked" IMO it should raise an exception, since that is what `mpz_as_uint_checked()` will do (if the mpz is negative).  If you don't want an exception then wouldn't `mp_obj_int_get_truncated()` work fine instead of using this function?

Jongy replied:
The intention of using this (in relation to the Linux kernel port) is mainly to catch errors in  arithmetic overflow errors before they propagate badly via the kernel FFI. From my personal experience those were usually overflows and not underflows and that's why I *think* I never encountered this negative small-int problem. That's also why I don't want to use `mp_obj_int_get_truncated`.

So yeah, it should raise an exception. I'll fix it.

Jongy replied:
Fixed.

Jongy commented on py/objint.c:
```diff
@@ -363,6 +363,10 @@ mp_int_t mp_obj_int_get_checked(mp_const_obj_t self_in) {
 }
 
+mp_uint_t mp_obj_int_get_uint_checked(mp_const_obj_t self_in) {
+    return MP_OBJ_SMALL_INT_VALUE(self_in);
```
Reopening the discussion of https://github.com/micropython/micropython/pull/5482#discussion_r363202922 here.

Jongy replied:
Removed this function.

Jongy commented:
I've rebased and fixed the negative small-int case for `mpz`. I split the entire function out because now keeping the logic for `int`/`uint` in the same function doesn't give that much of code-size benefit and has greater performance implications.

Regarding `longlong` and `none` impl - I wonder if it's even relevant to support them with this function?
`mp_obj_int_get_uint_checked` is currently an optional API that none of the builds are using. Builds that will use it (e.g Linux kernel) need it because they mess with unsigned machine words and they want protection from overflows. They will never use `none`/`longlong` for this case because both of them don't allow unsigned machine-word sized integers.

What do you think?

dpgeorge commented on py/objint_mpz.c:
```diff
@@ -411,6 +411,24 @@ mp_int_t mp_obj_int_get_checked(mp_const_obj_t self_in) {
+        if (mpz_as_uint_checked(&self->mpz, &value)) {
+            return value;
+        } else {
+overflow:
```
To avoid labels and goto (which is nice if possible) this function could be easily restructured so that it falls through to the bottom if there's an overflow, and the `mp_raise_msg` is at the bottom.  Please try to do that.

Jongy replied:
You are absolutely right. I dunno why I've structured it with a `goto` in the first place. I think it's a remnant of having a single function with `mp_obj_int_get_checked`.

dpgeorge commented on py/obj.h:
```diff
@@ -748,6 +748,8 @@ void mp_obj_cell_set(mp_obj_t self_in, mp_obj_t obj);
 mp_int_t mp_obj_int_get_truncated(mp_const_obj_t self_in);
 // Will raise exception if value doesn't fit into mp_int_t
 mp_int_t mp_obj_int_get_checked(mp_const_obj_t self_in);
+// Will raise exception if value doesn't fit into mp_uint_t
```
Maybe write "Will raise exception if value is negative or doesn't fit into mp_uint_t"

Jongy replied:
Fair

dpgeorge commented:
> I split the entire function out because now keeping the logic for `int`/`uint` in the same function doesn't give that much of code-size benefit and has greater performance implications.

+1



> Regarding `longlong` and `none` impl - I wonder if it's even relevant to support them with this function?

Let's not bother for now.  Can be done in the future if needed.

Jongy committed:
py/objint: Add mp_obj_int_get_uint_checked() helper.

Can be used where mp_obj_int_get_checked() will overflow due
to the sign-bit solely.
This returns an mp_uint_t, so it also verifies the given integer is
not negative.

Currently implemented only for mpz, since I don't expect this function
to be used by ports based on longlong/none impls.

Jongy commented:
> Let's not bother for now. Can be done in the future if needed.

Removed them, and fixed other comments.

dpgeorge commented:
Thanks for updating.  Merged in 176ab99180a95eeac8794828f2f751a696571bb5, with a coverage test added in 3448e69c2d6f6f066907005f56bbd26fffb756e9

Jongy commented:
Cool, I didn't know the `unix_coverage` build had this "unittest"-alike file.
